File: tools/jpf/jpf-symbc/src/examples/concolic/StatCalculator.java
package concolic;

import gov.nasa.jpf.symbc.Concrete;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import junit.framework.TestCase;

/**
 * This class serves as a way to calculate the median of a list of values.  It
 * is not threadsafe.
 */
public class StatCalculator implements Serializable
{
    static List values = new ArrayList();
    static double sum = 0;
    static double sumOfSquares = 0;
    static double mean = 0;
    static double deviation = 0;
    static int count = 0;
    
    public static void clear()
    {
        values.clear();
        sum = 0;
        sumOfSquares = 0;
        mean = 0;
        deviation = 0;
        count = 0;
    }

    public static void addValue(long newValue)
    {
        Number val = new Long(newValue);
        addValue(val);
    }

    public static void addValue(int newValue)
    {
    	Number val = new Integer(newValue);
        addValue(val);
        
    }

    public static void addValue(float newValue)
    {
        Number val = new Float(newValue);
        addValue(val);
    }

    public static void addValue(double newValue)
    {
        Number val = new Double(newValue);
        addValue(val);
    }

    public static Number getMedian()
    {
        return (Number) values.get(values.size() / 2);
    }
    
    public static double getMean()
    {
        return mean;
    }
    
    public static double getStandardDeviation()
    {
        return deviation;
    }
    
    public static Number getMin()
    {
        return (Number)values.get(0);
    }
    
    public static Number getMax()
    {
        return (Number)values.get(count-1);
    }
    
    public static int getCount()
    {
        return count;
    }

    public static void addValue(Number val)
    {
    //	if((double) val.intValue() > 10) {
    //		System.out.println("10");
    //	}
        int index = Collections.binarySearch(values, val);
        if (index >= 0 && index < values.size())
        {
            values.add(index, val);
        }
        else if (index == values.size() || values.size() == 0)
        {
            values.add(val);
        }
        else
        {;
            values.add((index * (-1)) - 1, val);
        }
        count++;
        System.out.println("stat ");
        double currentVal = val.doubleValue();
        sum += currentVal;
        sumOfSquares += currentVal * currentVal;
        mean = sum / count;
        deviation = Math.sqrt( (sumOfSquares / count) - (mean * mean) );
    }
    
}

File: experiments/src/programs.c#/StatTest/StatCalculator.cs
/*
 * Automatically converted from Java to C# using Sharpen.
 */

using System;
using System.Collections;
using Concolic;

namespace Concolic
{
	/// <summary>This class serves as a way to calculate the median of a list of values.</summary>
	/// <remarks>
	/// This class serves as a way to calculate the median of a list of values.  It
	/// is not threadsafe.
	/// </remarks>
	[System.Serializable]
	public class StatCalculator
	{
		internal static ArrayList values = new ArrayList();

		internal static double sum = 0;

		internal static double sumOfSquares = 0;

		internal static double mean = 0;

		internal static double deviation = 0;

		internal static int count = 0;

		public static void Clear()
		{
			values.Clear();
			sum = 0;
			sumOfSquares = 0;
			mean = 0;
			deviation = 0;
			count = 0;
		}


		public static double GetMedian()
		{
			return (double)values[values.Count / 2];
		}

		public static double GetMean()
		{
			return mean;
		}

		public static double GetStandardDeviation()
		{
			return deviation;
		}

		public static double GetMin()
		{
            return (double)values[0];
		}

		public static double GetMax()
		{
            return (double)values[count - 1];
		}

		public static int GetCount()
		{
			return count;
		}

		public static void AddValue(double val)
		{
			//	if((double) val.intValue() > 10) {
			//		System.out.println("10");
			//	}
			int index = values.BinarySearch(val);
			if (index >= 0 && index < values.Count)
			{
				values.Insert(index, val);
			}
			else
			{
				if (index == values.Count || values.Count == 0)
				{
					values.Add(val);
				}
				else
				{
					values.Insert((index * (-1)) - 1, val);
				}
			}
			count++;
			System.Console.Out.WriteLine("stat ");
			double currentVal = val;
			sum += currentVal;
			sumOfSquares += currentVal * currentVal;
			mean = sum / count;
			deviation = Math.Sqrt((sumOfSquares / count) - (mean * mean));
		}
	}
}
